src/pictureNamingTask.ts

```diff
@@ -0,0 +1,236 @@
+);
+let advancementSchedule = Number(experimentSettings.advancementSchedule);
+let regressionSchedule = Number(experimentSettings.regressionSchedule);
+// add logic to initialize with initialDifficulty
+// let intialDifficulty = Number(experimentSettings.initialDifficulty);
+let language = experimentSettings.language;
+let seed = experimentSettings.seed;
+
+console.log(
+  `Experiment will proceed with totalNumberOfTrialsToRun of ${totalNumberOfTrialsToRun}, an advancementSchedule of ${advancementSchedule}, and a regressionSchedule of ${regressionSchedule}`,
+);
+
+/*
+functions for generating
+experiment_stimuli
+*/
+
+const indiciesSelected = new Set();
+let rng = prand.xoroshiro128plus(seed);
+
+function getRandomElementWithSeed(array: any[]) {
+  let randomIndex: number;
+  let foundUnique = false;
+
+  // if all images have been shown clear the set
+  if (indiciesSelected.size === array.length) {
+    indiciesSelected.clear();
+  }
+
+  do {
+    const [newRandomIndex, newRng] = prand.uniformIntDistribution(
+      0,
+      array.length - 1,
+      rng,
+    );
+    rng = newRng;
+    randomIndex = newRandomIndex;
+
+    if (!indiciesSelected.has(randomIndex)) {
+      indiciesSelected.add(randomIndex);
+      foundUnique = true;
+    }
+  } while (!foundUnique);
+
+  console.log("Selected Index:", randomIndex);
+  console.log("Selected Indices Set:", indiciesSelected);
+
+  let result = [array[randomIndex]];
+  console.log("***********************");
+  console.log("Result from random index:");
+  console.table(result);
+  console.log("***********************");
+  return result;
+}
+
+// draw an image at random from the bank depending on the difficulty_level selected
+function createStimuli(
+  difficultyLevel: number,
+  language: string,
+  clearSet: boolean,
+) {
+  if (clearSet === true) {
+    indiciesSelected.clear();
+  }
+  try {
+    let imgList = imageDB;
+    imgList = imgList.filter(
+      (image: any) =>
+        Number(image.difficultyLevel) === difficultyLevel &&
+        image.language === language,
+    );
+    console.log("filtered");
+    console.log(imgList);
+    let result = getRandomElementWithSeed(imgList);
+    return result;
+  } catch (error) {
+    console.error("Error fetching and parsing data:", error);
+  }
+}
+
+//****************************
+//********EXPERIMENT**********
+//****************************
+// a timeline is a set of trials
+// a trial is a single object eg htmlKeyboardResponse etc ...
+const timeline: any[] = [];
+export default function pictureNamingTask(difficultyLevelParam: number) {
+  let experiment_stimuli = createStimuli(difficultyLevelParam, language, false);
+  let currentDifficultyLevel: number = difficultyLevelParam;
+  if (difficultyLevelParam) {
+    const jsPsych = initJsPsych({
+      on_finish: function () {
+        transformAndDownload(jsPsych.data);
+      },
+    });
+
+    const welcome = {
+      type: HtmlKeyboardResponsePlugin,
+      stimulus: `${welcomeText}`,
```

Comment:
what is the point of the template literal here?

Reply:
none, removed